### ISSUE TITLE
Adding Transformer/CNN model

### ISSUE THREAD

**Issue by patelrajnath** [ModernMT/MMT#377]
Are there planes to add Transformer/CNN model in MMT?

**Comment by davidecaroselli** [comment]
Hi @patelrajnath 

the transformer has been integrated in the paid Enterprise Edition of ModernMT that will be release very soon. There are no plans for the moment to port this integration in the open source, but this can change in the future.

We will constantly update the Open Source version of course, but for the moment we will keep the RNN model.

**Comment by mzeidhassan** [comment]
@davidecaroselli  I hope you don't mind sharing my view  on this with you here, Davide! I hope you consider adding Transformer to the open-source version. As far as I know, MMT is using OpenNMT-py which integrates Transformer already. Since Transform seems to be the standard now in most open-source NMT solutions, I am afraid MMT will lose its user-base if this is not implemented and users will tend to use other solutions.

**Comment by davidecaroselli** [comment]
Hi @mzeidhassan 

we are really aware of the problem, and our decision is not definitive and it can change in the future: we will for sure release the transformer model eventually.

As for the OpenNMT-py Transformer: I have tried twice in several months to use their transformer and both times I found severe problems in that implementation. I was not able to get reasonable results due to (I suppose) some problems in the implementation or in the configuration. For this reason we decided not to include it in the Open Source: at the end it was performing much worse than the current  RNN implementation.

I don't know if things are solved now, but if you search inside OpenNMT-py issues page you'll find all my issues and discussions.

Cheers,
Davide

**Comment by davidecaroselli** [comment]
@mzeidhassan @patelrajnath 

I have good news. We have an internal discussion on this topic, and we found the right compromise for us that will allow us publish a **Transformer release for ModernMT**. We will first focus on completing the implementation internally and we will then publish the 3.0 release with tensor2tensor integration with Transformer.

*Forgive me if I hide our previous conversation*, but given this important decision, I don't want to confuse future readers (I hope you'll understand :) )

I'll keep you updated on this thread.

Best,
Davide

**Comment by patelrajnath** [comment]
That's a great news. Thanks. :)

**Comment by mzeidhassan** [comment]
Thank you so much @davidecaroselli for being always supportive and flexible. I believe this is the right way to go. Thanks a million!

**Comment by davidecaroselli** [comment]
Hi @patelrajnath and @mzeidhassan 

it's finally here! On the branch **release/v3.0** there is a working implementation of ModernMT with tensor2tensor Transformer.

It is still in an early stage of development, do we will probably change something and improve some parts but... it's working and you can start playing with it right away!

Enjoy.

**Comment by patelrajnath** [comment]
@davidecaroselli 

Great! Thanks for making it available. :+1:

**Comment by LoekvanKooten** [comment]
@davidecaroselli There's no nvidia-docker image of this yet? sudo nvidia-docker pull modernmt/release/3.0 gives "access denies, repository does not exist or may require docker login".

**Comment by davidecaroselli** [comment]
Hi @LoekvanKooten 

no, as I mentioned, this is still "raw" code - not a well packaged release. But you can still install it and play with it by Installing ModernMT from source (plus installing Tensorflow lib).

We will soon update code, releases and all Install documentation in order to match this new integration but this is still not a release-ready code and it will definitely change in the near future.

**Comment by LoekvanKooten** [comment]
A really noob question: if I use the standard install procedure as descrived in Option 3, does that install release/v3.0 or master? Probably, master, right? As Install.md hasn't been updated yet, what needs to be changed to install 3.0 instead? (I should add that I successfully built the source - I just don't know which one it is).

Second question: do the engines need to be rebuilt for the Transformer to kick in?

**Comment by davidecaroselli** [comment]
Hi @LoekvanKooten 

yeah, Install.md is still not ready to explain how to install `release/v3.0`. However, quick way to test it would be:
1. Don't install pyTorch (just skip that bit)
2. Install tensorflow following its [guide for Ubuntu](https://www.tensorflow.org/install/install_linux)
3. Right after the `git clone` type `git checkout release/v3.0`

We will update the documentation as soon as the code will be more stable to put it in master branch. 
You need to re-train the engine because the architecture is completely different from previous versions (T2T Transformer vs. OpenNMT seq2seq).

I cannot assure you that you'll be able to use the same engine you're about to train also with the actual v3.0 release, as I said the code will probably change. So please, don't use it now for production system  :)

Best,
Davide

**Comment by davidecaroselli** [comment]
Just run "python download_dependencies.py" instead.

**Comment by LoekvanKooten** [comment]
That worked! However, now ",/mmt create ja nl data/japdut -e japdut --neural" gives "unrecognized argument: --neural". ./mmt status says there are no engines (which is correct, so it can run).

**Comment by davidecaroselli** [comment]
No need to specify "--neural" anymore: neural is the only option available in v3.0

**Comment by LoekvanKooten** [comment]
Aha, that explains a lot.

Corpora cleaning: Unexpected exception: sequence item 0: expected string, NoneType found.

This is exactly the same corpora that works in the master branch. All files neatly start with an actual string.  This is for Japanese-Dutch.

**Comment by davidecaroselli** [comment]
@LoekvanKooten update the code (simply run `git pull origin release/v3.0`) and retry. This time you should see a more detailed exception with a stack trace. This will help trying to understand where the problem is.

**Comment by LoekvanKooten** [comment]
Done!

`Traceback (most recent call last):
  File "./mmt", line 605, in <module>
    main()
  File "./mmt", line 551, in main
    actions[command](args)
  File "./mmt", line 143, in main_create
    builder.build()
  File "/home/loek/MMT/cli/mmt/engine.py", line 529, in build
    self._build(resume=False)
  File "/home/loek/MMT/cli/mmt/engine.py", line 602, in _build
    method(self, args, skip=skip, log=log_stream, delete_on_exit=self._delete_on_exit)
  File "/home/loek/MMT/cli/mmt/engine.py", line 401, in __call__
    self._f(*args, **kwargs)
  File "/home/loek/MMT/cli/mmt/engine.py", line 676, in _clean_tms
    args.corpora = self._cleaner.clean(args.corpora, folder, log=log)
  File "/home/loek/MMT/cli/mmt/engine.py", line 41, in clean
    command = mmt_javamain(self._java_main, args=args, max_heap_mb=extended_heap_mb)
  File "/home/loek/MMT/cli/__init__.py", line 44, in mmt_javamain
    classpath = ':'.join(classpath)
TypeError: sequence item 0: expected string, NoneType found
`

**Comment by davidecaroselli** [comment]
@LoekvanKooten have you run `mvn clean install` as stated in Install.md instructions? It seems you don't have the `mmt-2.5.jar` file in your `build` folder.

**Comment by LoekvanKooten** [comment]
I'm an idiot. Thank you!

**Comment by LoekvanKooten** [comment]
```
INFO: (4 of 6) Preparing data...                                      
ERROR Unexpected exception: 

Traceback (most recent call last):
  File "./mmt", line 605, in <module>
    main()
  File "./mmt", line 551, in main
    actions[command](args)
  File "./mmt", line 143, in main_create
    builder.build()
  File "/home/loek/MMT/cli/mmt/engine.py", line 529, in build
    self._build(resume=False)
  File "/home/loek/MMT/cli/mmt/engine.py", line 602, in _build
    method(self, args, skip=skip, log=log_stream, delete_on_exit=self._delete_on_exit)
  File "/home/loek/MMT/cli/mmt/engine.py", line 401, in __call__
    self._f(*args, **kwargs)
  File "/home/loek/MMT/cli/mmt/engine.py", line 734, in _prepare_data
    log=log, bpe_symbols=self._bpe_symbols)
  File "/home/loek/MMT/cli/mmt/engine.py", line 155, in prepare_data
    osutils.shell_exec(command, stdout=log, stderr=log, env=env)
  File "/home/loek/MMT/cli/libs/osutils.py", line 41, in shell_exec
    shell=(True if isinstance(cmd, basestring) else False), env=env)
  File "/usr/lib/python2.7/subprocess.py", line 711, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1343, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

**Comment by davidecaroselli** [comment]
ahahaha no please, this kind of things just happens!

On this second problem: it seems you don't have `t2t-datagen` available on the shell (and I bet also `t2t-trainer`).

Depending on which installation option you picked from tensorflow page, you should enable your conda/vistualenv python environment before running ModernMT. If you type `t2t-datagen` on your shell and it says "command not found" then neither MMT will work.

**Comment by LoekvanKooten** [comment]
You're too kind!

I assume that running tensorflow in a separate docker won't work then at this stage?

I installed Anaconda (working), created a tensorflow environment in it (working), but then the first test fails:

`pip install --ignore-installed --upgrade https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.9.0-cp27-none-linux_x86_64.whl

(tensorflow) loek@DORAEMON:~$ python
Python 2.7.15 |Anaconda, Inc.| (default, May  1 2018, 23:32:55) 
[GCC 7.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> # Python
... import tensorflow as tf
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/loek/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/__init__.py", line 22, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "/home/loek/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/home/loek/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow.py", line 74, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "/home/loek/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/home/loek/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/home/loek/anaconda2/envs/tensorflow/lib/python2.7/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
ImportError: libcublas.so.9.0: cannot open shared object file: No such file or directory


Failed to load the native TensorFlow runtime.
`

**Comment by LoekvanKooten** [comment]
Solved! The installation instructions for Tensorflow are wrong. By default CUDA 9.1 is installed, but the newest version of tensorflow is looking for 9.0 libraries. This can be solved by installing CUDA 9.0 too.

**Comment by LoekvanKooten** [comment]
O, and installing Tensor2Tensor helps. This is a separate install from Tensorflow itself, which may not be immediately clear for n00bs like me. And of course it needs to be installed with pip *inside* the tensorflow virtual environment.

**Comment by davidecaroselli** [comment]
I promise, it will be all much much simpler once I completed the integration **and** the documentation 😄

**Comment by LoekvanKooten** [comment]
@davidecaroselli 

It's okay! I'm learning so much and you are very patient. One last question I think: any idea where tensorflow keeps its logs? I'm trying to fire up Tensorboard with --logdir, but I don't know the --logdir and feeding MMT/runtime/<engine name>/logs obviously doesn't do much.

**Comment by davidecaroselli** [comment]
Yes!
Assuming you're training the default engine (no `-e` option in `mmt create`) they should be in `runtime/default/tmp/training/neural_model`

**Comment by LoekvanKooten** [comment]
@davidecaroselli 

That worked! I remember running Tensorflow on its own a long time ago, and then it actually showed words in the Projector, but now I see only numbers. Is this due to the inherent workings of MMT?

**Comment by LoekvanKooten** [comment]
@davidecaroselli

I can confirm that the baseline of MMT on steroids (tensorflow) performs considerably better than the old version on the same dataset (after an arbitrary stop during training, and without updates as that doesn't seem to be working yet). It's not absolutely shocking, but the translations read more smoothly and natural.

**Comment by mzeidhassan** [comment]
@davidecaroselli This is the best news ever to hear after my vacation :-) Thank you so much Davide for making this available in MMT 3.0. Cannot wait for the stable release. Thank again!

**Comment by LoekvanKooten** [comment]
@davidecaroselli 

I can also confirm that ModernMT beats the competition in every possible way, DeepL and Google Translate included. Of course a major factor is that this is one of the first neural MT programs you can feed with your own data, but what makes it even more interesting is that you can update it on the fly. Average Levenshtein scores for Google/DeepL hover around 22. ModernMT hovers around 11. The adaptive function makes sure that accuracy score raises about 2% per 10,000 translated words (from 67% to 69%). Just some numbers I got. 

As you may know I have always been very critical towards MT, but this is the first software that can actually work with instead of work against the translator.

I'm curious though what Google will do now you can feed it your own data (probably not on the fly though)?

I will write a full report in a few weeks.

**Comment by mzeidhassan** [comment]
Thanks @LoekvanKooten for sharing your experience. This is great news to hear. 

@davidecaroselli any rough date when the 3.0 stable version will be available? I cannot wait to test this. Thanks a million Davide for making this available.

**Comment by davidecaroselli** [comment]
@mzeidhassan I guess it will take less than a week to complete. However we will try to run as many tests as possible given the whole new core engine. I hope it will be ready to be published in a couple of weeks!

**Comment by davidecaroselli** [comment]
Hi all,

we have finally released [ModernMT v3.0 with T2T Transformer Model](https://github.com/ModernMT/MMT/releases/tag/v3.0): thank you all for the patience :)

@cettolo 
> is the "averaging step" named "Pack model" in MMT output (the 6th step out of the 6 training steps)?

Yes, the main activity during the 6th step is the averaging of the model

> how many ckpt's are averaged?

All the checkpoints available in the training folder. Namely 20, but due a "weird" behaviour of T2T  you only get 10 (all checkpoints are saved twice, so you have 1000, 1001, 2000, 2001 and so on).

> is that a number that can be somehow passed as an option?

Not at the moment. However if run the training with `--debug` flag, the process will keep all the checkpoints in `runtime/tmp/training` folder. You can access the last 20 (or 10) checkpoints and run the average checkpoints manually if you wish to make some test.

> is it possible to run a mmt command to perform the average? 

If the training terminates unexpectedly, you can always resume it by adding `--resume` to the training command you just run. However to run manual average you can use the official script of T2T: [avg_checkpoints.py](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/utils/avg_checkpoints.py) 

> why did you dismiss the phrase-based decoder in v3.0? 

Phrase-based decoder is no longer supported and won't be updated. For this reason we decided to remove it from the v3.0 and all the up-coming releases. This does not mean that PB-decoder is not available anymore: the last stable phrase-based release is 2.5 and you will aways be able to download it and install it in the release page: [ModernMT v2.5](https://github.com/ModernMT/MMT/releases/tag/v2.5)

Thank you all for your support, I'm closing this issue because the v3.0 is not ready and available. If you have any other question please don't hesitate to re-open this issue, or start a new thread for specific issues/discussions.

Best,
Davide
